Initial commit

shadyvb opened the pull request:
This plugin hooks into the file upload process and encrypts files when a checkbox in the upload UI is checked. And filters the attachment URL to provide a real-time decrypted version behind a capability check.

shadyvb committed:
Initial commit

shadyvb committed:
Fix namespace and plugin meta

shadyvb committed:
Add a plugin activation hook to flush rewrite rules

goldenapples commented on namespace.php:
```diff
@@ -0,0 +1,247 @@
+add_action( 'post-upload-ui', __NAMESPACE__ . '\\post_upload_ui_encryption', 10 );
+
+/**
+ * Adds the Javasript handling of the encryption checkbox / params
```
typo: javascript

goldenapples commented on namespace.php:
```diff
@@ -0,0 +1,247 @@
+	 * @return bool
+	 */
+	if ( ! apply_filters( 'encrypted_uploads_should_encrypt', true, $ext_info['type'], $details['file'] ) ) {
+		$details['error'] = esc_html__( 'Could not encrypt the file, possibly because of filesystem permissions.', 'encrypted-uploads' );
```
This doesn't look like the right error message here...?

goldenapples reviewed (approved):
This looks good to me, and functions as expected.  A couple notes:

The docblocks could use some cleaning up throughout for clarity and style: summaries should be written in sentence format ending with a full stop, param types should be documented more fully.

I'm wondering if the upload encryption could be done with a filter on `wp_handle_upload_prefilter` instead of waiting until the upload has been saved to disk, and then replaced with an encrypted version. (just a question, not a suggestion, since I don't know the answer...) As it stands, if there's an error encrypting a file for any reason, the upload will be available unencrypted, which seems like an undesirable failure mode.

shadyvb committed:
Tweak docblocks, add file extension param to encryption check filter

shadyvb committed:
Fix wrong errorr message

shadyvb commented:
@goldenapples Thanks for the insights!

I guess I opted for that filter so WordPress would carry out all the checks instead of me duplicating everything. That's where VIP Go hooks and I think it's best for this use.

And regarding files being uploaded with no decryption in case of failure, that's definitely a good concern, can you point me to how this might happen ? I thought I handled that with returning `$details` with an `'error'` property.

goldenapples commented on namespace.php:
```diff
@@ -0,0 +1,253 @@
+		throw new Exception( esc_html__( 'openssl_encrypt function does not exist, cannot proceed with encryption' ) );
+	}
+
+	$encrypted = openssl_encrypt( $contents, 'AES128', ENCRYPTED_UPLOADS_CIPHER_KEY );
```
Should 'AES128' here be `ENCRYPTED_UPLOADS_CIPHER_METHOD`?

goldenapples commented:
> regarding files being uploaded with no decryption in case of failure, that's definitely a good concern, can you point me to how this might happen ? I thought I handled that with returning $details with an 'error' property.

I guess that's not exactly what my concern was. There are a few cases where `encrypt_file()` could fail or error: the site has an invalid cypher defined in `ENCRYPTED_UPLOADS_CIPHER_METHOD`, the `fopen` permissions check fails, an out of memory error occurs... If one of those things happens, then the attachment file will remain in the uploads directory unprotected.

I thought it might be more secure to hook in before the file is saved, so that the unencrypted upload is never saved to disk. But without knowing the desired use case, its hard to say what level of protection is necessary.

shadyvb committed:
Fix hardcoded param

shadyvb commented:
~@nathanielks~ @goldenapples  Thanks for explaining, the file will definitely linger there until the end of time. Good thinking. Thinking `pre_move_uploaded_file` might be a better fit, will have a go at it 👍 .

goldenapples commented:
> Thinking pre_move_uploaded_file might be a better fit, will have a go at it.

Oh yeah, that does look like the best place to hook into there....

shadyvb committed:
Use %_prefilter instead of wp_handle_upload to stop file from being saved to uploads if failed

shadyvb commented:
@goldenapples used `%_prefilter` instead as `pre_move_uploaded_file` is not supposed to be a checkpoint, rather than an override to actually move the file manually, `$file['error']` is not checked afterwards.

Thanks for the insights!

goldenapples commented:
This looks :100: to me.

I have two questions about the use of `openssl_encrypt()` though.

* By default the encrypted data is being saved in base64 format. This makes sense for the post meta value being stored on the attachment post, because anything else would risk corruption when it was saved to the database. But the upload file itself is saved as binary, so this is unnecessary and a bit wasteful. If you pass `OPENSSL_RAW_DATA` as the fourth parameter to `openssl_encrypt()` (and `openssl_decrypt()`), you'll get to file sizes about 30% smaller.

* I'm seeing a warning `Warning: openssl_encrypt(): Using an empty Initialization Vector (iv) is potentially insecure and not recommended` which relates to the fifth argument for `openssl_encrypt()`. As I'm understanding, the IV is like a unique salt which is used in encrypting the file to ensure that if the same file is uploaded twice, each one gets a unique encrypted signature. This prevents an user who has access to the encrypted uploads from identifying an encrypted file by matching it to the encryption output of a known file.

Would it be useful to add an initialization vector to the encryption process? I think that the 'encrypted-upload' meta field, trimmed to the appropriate length, could probably be used for this purpose.

shadyvb committed:
Proper handling of openssl options and iv. props @goldenapples

shadyvb committed:
Use a WP salt constant as an IV instead, as a filename is not a proper IV

shadyvb commented:
Thanks for the tips @goldenapples, very valuable!

I think using an existing WP Salt constant makes sense for an IV ? As the filename / ID might still be exposed / guessed easily. While a salt key should fundamentally be unique and secure.

Let me know what you think about the changes.

goldenapples commented:
> I think using an existing WP Salt constant makes sense for an IV ? As the filename / ID might still be exposed / guessed easily. While a salt key should fundamentally be unique and secure.

I don't think so; I think using the WP salt as a fallback for the _cipher key_, as you're doing here, makes good sense. That provides security for the files, so its not possible to decrypt them without access to that salt/cipher key.

The iv protects against a different kind of attack, and it needs to be unique for each file being encrypted. [Here's a basic SO explanation that helped me understand it.](https://stackoverflow.com/questions/11821195/use-of-initialization-vector-in-openssl-encrypt#12486940) Basically, it assumes that the files can't be decrypted, but that a user can check to see if another user has uploaded a specific file by uploading it themselves and comparing the encrypted version of their files to all the other encrypted files in the uploads directory. If each file has a unique iv, then the encrypted versions won't match.

shadyvb committed:
Add a site-unique IV salt, and generate random salt for each file

shadyvb commented:
@goldenapples I've spiced it a bit more by adding a site-specific generated-once random IV salt, and a file-unique IV as well.

Let me know what you think.

goldenapples commented:
Code looks good to me. I like the idea of writing the initialization vector for each file directly to the beginning of the file - that's an elegant way of making it easy to retrieve while still being unique and cryptographically secure.

Are there any character-set or collation concerns with saving the `openssl_random_pseudo_bytes()` unencoded to the database? I can't find any issues, but I'm always wary of saving binary data in varchar mysql fields without base64 or hex encoding first. Especially since if that option were to be deleted or changed for any reason, all the files would become unreadable. 

I'm also getting some issues in my local environment where I can't view encypted files after uploading them. I haven't figured out what the problem is yet, though. Does this code work for you?

While testing, I also noticed one more UI issue. If the "Encrypt" checkbox is checked, and you drag and drop a file onto the uploader, it won't be encrypted when uploading.

Otherwise, I think this is great.

shadyvb committed:
base64_encode the rand string just to be safe

`openssl rand` does accept a flag to base64_encode the result, so just being safe here, as this renders the file impossible to decode if lost / changed.

shadyvb committed:
Force boolean check of the get param

shadyvb committed:
base64 encode the post id key to prevent cases leading to invalid URLs

shadyvb commented:
- So I've checked and found the `openssl rand` command to enable base64 encoding of the result, so to be safe I'll just encode the site salt which is meant to be stored in DB.
- The bug you found trying to decode some files might just be that you uploaded the same file, and the get call was just cached and you hit a previously encrypted file using an older IV. This shouldn't happen in real-life.
- The encoding of the post ID as well can break the URL, if it has leading slashes, so I'm base64_encoding that as well now.

shadyvb committed:
Do not cache decrypted files

shadyvb committed:
Code alignment / line reorder
